src/combinatorics/hall/basic.lean

```diff
@@ -0,0 +1,230 @@
+{ obj := λ ι', hall_matchings_on t ι'.unop,
+  map := λ ι' ι'' g f, hall_matchings_on.restrict t (category_theory.le_of_hom g.unop) f }
+
+noncomputable instance hall_matchings_on.fintype {ι α : Type*}
+  (t : ι → finset α) (ι' : finset ι) :
+  fintype (hall_matchings_on t ι') :=
+begin
+  classical,
+  rw hall_matchings_on,
+  let g : hall_matchings_on t ι' → (ι' → ι'.bUnion t),
+  { rintro f i,
+    refine ⟨f.val i, _⟩,
+    rw mem_bUnion,
+    exact ⟨i, i.property, f.property.2 i⟩ },
+  apply fintype.of_injective g,
+  intros f f' h,
+  simp only [g, function.funext_iff, subtype.val_eq_coe] at h,
+  ext a,
+  exact h a,
+end
+
+end hall_marriage_theorem
+
+open hall_marriage_theorem
+
+/--
+This is the version of **Hall's Marriage Theorem** in terms of indexed
+families of finite sets `t : ι → finset α`.  It states that there is a
+set of distinct representatives if and only if every union of `k` of the
+sets has at least `k` elements.
+
+Recall that `s.bUnion t` is the union of all the sets `t i` for `i ∈ s`.
+
+This theorem is bootstrapped from `finset.all_card_le_bUnion_card_iff_exists_injective'`,
+which has the additional constraint that `ι` is a `fintype`.
+-/
+theorem finset.all_card_le_bUnion_card_iff_exists_injective
+  {ι α : Type*} [decidable_eq α] (t : ι → finset α) :
+  (∀ (s : finset ι), s.card ≤ (s.bUnion t).card) ↔
+    (∃ (f : ι → α), function.injective f ∧ ∀ x, f x ∈ t x) :=
+begin
+  split,
+  { intro h,
+    haveI : ∀ (ι' : (finset ι)ᵒᵖ), nonempty ((hall_matchings_functor t).obj ι') :=
+      λ ι', hall_matchings_on.nonempty t h ι'.unop,
+    classical,
+    haveI : Π (ι' : (finset ι)ᵒᵖ), fintype ((hall_matchings_functor t).obj ι') := begin
+      intro ι',
+      rw [hall_matchings_functor],
+      apply_instance,
+    end,
+    obtain ⟨u, hu⟩ := nonempty_sections_of_fintype_inverse_system (hall_matchings_functor t),
+    refine ⟨_, _, _⟩,
+    { exact λ i, (u (opposite.op ({i} : finset ι))).val
+                 ⟨i, by simp only [opposite.unop_op, mem_singleton]⟩, },
+    { intros i i',
+      have subi : ({i} : finset ι) ⊆ {i,i'} := by simp,
+      have subi' : ({i'} : finset ι) ⊆ {i,i'} := by simp,
+      have le : ∀ {s t : finset ι}, s ⊆ t → s ≤ t := λ _ _ h, h,
+      rw [←hu (category_theory.hom_of_le (le subi)).op,
+          ←hu (category_theory.hom_of_le (le subi')).op],
+      let uii' := u (opposite.op ({i,i'} : finset ι)),
+      exact λ h, subtype.mk_eq_mk.mp (uii'.property.1 h), },
+    { intro i,
+      apply (u (opposite.op ({i} : finset ι))).property.2, }, },
+  { rintro ⟨f, hf₁, hf₂⟩ s,
+    rw ←finset.card_image_of_injective s hf₁,
+    apply finset.card_le_of_subset,
+    intro _,
+    rw [finset.mem_image, finset.mem_bUnion],
+    rintros ⟨x, hx, rfl⟩,
+    exact ⟨x, hx, hf₂ x⟩, },
+end
+
+/-- Given a relation such that the image of every singleton set is finite, then the image of every
+finite set is finite. -/
+instance {α β : Type*} [decidable_eq β]
```

Comment:
Could this go in an earlier file? I don't think `fintype` and `rel` are both imported before this one, so maybe at least put it at the top of this file?

Reply:
I put it right before the thing that actually uses it.  I don't think it makes sense going at the top of this file, but it certainly would make sense to put it in a module about relations.  The problem with `data.rel` is that it doesn't import anything with fintypes.

The last time this was code reviewed, we left moving it for the future.  In this PR, I've left it in its same relative position.